### PR TITLE
7018932 : Drawing very large coordinates with a dashed Stroke can cause Java to hang

### PR DESCRIPTION
**Pull request by bourgesl**
This is my fix proposal to this bug:
- added new method strokeTo(... Region clip ...) in the abstract RenderingEngine class
- fixed all RenderingEngine implementations in java.desktop module
- MarlinRenderingEngine now uses the clip region to roughly & quickly clip the given shape in strokeTo(clip) 
- LoopPipe.getStrokeSpans() uses the new strokeTo(clip) method to get good performance with huge dashed shapes.

I wrote a new test class to validate the bug fix.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-7018932](https://bugs.openjdk.java.net/browse/JDK-7018932): Drawing very large coordinates with a dashed Stroke can cause Java to hang

### Reviewers
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**)
 * [Phil Race](https://openjdk.java.net/census#prr) (@prrace - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2013/head:pull/2013`
`$ git checkout pull/2013`

### CODE REVIEW

File: src/java.desktop/share/classes/sun/java2d/pipe/RenderingEngine.java [hunk 1]
```diff
@@ -218,6 +222,47 @@ public abstract void strokeTo(Shape src,
...
+     *             than the minimum dropout pen width
+     * @param normalize indicates whether stroke normalization should
+     *                  be applied
+     * @param antialias indicates whether or not adjustments appropriate
+     *                  to antialiased rendering should be applied
+     * @param consumer the {@code PathConsumer2D} instance to forward
+     *                 the widened geometry to
+     * @since 17
+     */
+    public abstract void strokeTo(Shape src,
```
Comment (mrserb) [inline src/java.desktop/share/classes/sun/java2d/pipe/RenderingEngine.java:257]: I think we need to clarify the backward compatibility of this change.
 - If we would like to support the old marlin version via "-Dsun.java2d.renderer" option in jdk17 then adding the new abstract method to this interface breaks this possibility. To support that this method should be made default and call the old strokeTo method.
 - If support for the old marlin library is not necessary then we could drop the old method from this interface(but preserve it in the MarlinRenderingEngine to have the same marlin codebase), it won't be used anyway.

File: src/java.desktop/share/classes/sun/java2d/pipe/LoopPipe.java [hunk 2]
```diff
@@ -261,17 +261,25 @@ public static ShapeSpanIterator getStrokeSpans(SunGraphics2D sg2d,
...
+            final Region clip = sg2d.getCompClip();
+            sr.setOutputArea(clip);
             sr.setRule(PathIterator.WIND_NON_ZERO);
 
             BasicStroke bs = (BasicStroke) sg2d.stroke;
             boolean thin = (sg2d.strokeState <= SunGraphics2D.STROKE_THINDASHED);
             boolean normalize =
                 (sg2d.strokeHint != SunHints.INTVAL_STROKE_PURE);
 
+// No clipping (pre-jdk17)
```
Comment (mrserb) [inline src/java.desktop/share/classes/sun/java2d/pipe/LoopPipe.java:273]: This code is not a part of the marlin, so exists only in the jdk17. Do we need these commented lines here?

File: test/jdk/sun/java2d/marlin/DrawingTest7018932.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import javax.swing.JFrame;
+import javax.swing.JPanel;
+
+/*
+ * @test
+ * @bug 7018932
+ * @summary fix LoopPipe.getStrokedSpans() performance (clipping enabled by Marlin renderer)
+ * @run main DrawingTest7018932
+ */
+public class DrawingTest7018932 extends JPanel {
```
Comment (mrserb) [inline test/jdk/sun/java2d/marlin/DrawingTest7018932.java:42]: This will work only in the headful environment, I suggest something like this:
```
/*
 * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

import java.awt.BasicStroke;
import java.awt.Graphics2D;
import java.awt.RenderingHints;
import java.awt.Stroke;
import java.awt.geom.Line2D;
import java.awt.image.BufferedImage;

import static java.awt.image.BufferedImage.TYPE_INT_ARGB_PRE;

/*
 * @test
 * @bug 7018932
 * @summary fix LoopPipe.getStrokedSpans() performance (clipping enabled by Marlin renderer)
 */
public final class StrokedLinePerf {

    public static void main(String[] args) {
        BufferedImage bi = new BufferedImage(400, 400, TYPE_INT_ARGB_PRE);
        test(bi, true);
        test(bi, false);
    }

    private static void test(BufferedImage bi, boolean useAA) {
        Graphics2D g2d = bi.createGraphics();
        long start = System.nanoTime();

        if (useAA) {
            g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING,
                    RenderingHints.VALUE_ANTIALIAS_ON);
        } else {
            g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING,
                    RenderingHints.VALUE_ANTIALIAS_OFF);
        }

        Stroke stroke = new BasicStroke(2.0f, 1, 0, 1.0f, new float[]{0.0f, 4.0f}, 0.0f);
        g2d.setStroke(stroke);

        //Large values to trigger crash / infinite loop.
        g2d.draw(new Line2D.Double(4.0, 1.794369841E9, 567.0, -2.147483648E9));

        System.out.println("Test duration= " + (1e-6 * (System.nanoTime() - start)) + " ms.");
        g2d.dispose();
    }
}
```

File: src/java.desktop/share/classes/sun/java2d/pipe/RenderingEngine.java (hunk 1 above)
Comment (bourgesl) [reply src/java.desktop/share/classes/sun/java2d/pipe/RenderingEngine.java:257]: I like the first proposal: use default class method as RE is an abstract class.

File: src/java.desktop/share/classes/sun/java2d/pipe/LoopPipe.java (hunk 2 above)
Comment (bourgesl) [reply src/java.desktop/share/classes/sun/java2d/pipe/LoopPipe.java:273]: Will remove the comment

File: test/jdk/sun/java2d/marlin/DrawingTest7018932.java (hunk 3 above)
Comment (bourgesl) [reply test/jdk/sun/java2d/marlin/DrawingTest7018932.java:42]: Interesting, but BufferedImage rendering may use another java2d pipeline... so I will check if LoopPipe.getStrokedSpans() is used by your test code.

Comment (bourgesl) [reply test/jdk/sun/java2d/marlin/DrawingTest7018932.java:42]: Moreover the bug was detected on both gdi & xrender native pipelines...

Comment (bourgesl) [reply test/jdk/sun/java2d/marlin/DrawingTest7018932.java]: Confirmed:
* useAA=true:
sun.java2d.marlin.MarlinRenderingEngine.getAATileGenerator(java.awt.geom.Line2D$Double, AffineTransform[[1.0, 0.0, 0.0], [0.0, 1.0, 0.0]], Region[[0, 0 => 400, 400]], java.awt.BasicStroke@110e1781, wide, normalized)
* useAA=false:
sun.java2d.marlin.MarlinRenderingEngine.strokeTo(java.awt.geom.Line2D$Double, AffineTransform[[1.0, 0.0, 0.0], [0.0, 1.0, 0.0]], Region[[0, 0 => 400, 400]], java.awt.BasicStroke@110e1781, wide, normalized, non-AA, sun.java2d.pipe.ShapeSpanIterator)

File: src/java.desktop/share/classes/sun/java2d/marlin/TransformingPathConsumer2D.java [hunk 4]
```diff
@@ -898,22 +899,22 @@ public long getNativeConsumer() {
         final float[] clipRect;
 
         // clip rectangle (ymin, ymax, xmin, xmax) including padding:
-        final float[] clipRectPad = new float[4];
+        final double[] clipRectPad = new double[4];
```
Comment (mrserb) [inline src/java.desktop/share/classes/sun/java2d/marlin/TransformingPathConsumer2D.java:902]: It does not seem related to this fix? Or an updated test uncovered this?

File: test/jdk/sun/java2d/marlin/StrokedLinePerf.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import java.awt.geom.Line2D;
+import java.awt.image.BufferedImage;
+
+import static java.awt.image.BufferedImage.TYPE_INT_ARGB_PRE;
+
+/**
+ * @test
+ * @bug 7018932
+ * @summary fix LoopPipe.getStrokedSpans() performance (clipping enabled by Marlin renderer)
+ * @run main DrawingTest7018932
```
Comment (mrserb) [inline test/jdk/sun/java2d/marlin/StrokedLinePerf.java:37]: typo DrawingTest7018932 -> StrokedLinePerf

Comment (bourgesl) [reply test/jdk/sun/java2d/marlin/StrokedLinePerf.java]: Removed invalid test

File: src/java.desktop/share/classes/sun/java2d/marlin/TransformingPathConsumer2D.java (hunk 4 above)
Comment (bourgesl) [reply src/java.desktop/share/classes/sun/java2d/marlin/TransformingPathConsumer2D.java:902]: This fix in Marlin (float) renderer is needed as this numerical precision issue was triggered by the StrokedLinePerf using the MarlinRenderingEngine. It is very similar to DTransformingPathConsumer2D (but reuse was not trivial).

I propose to remove the MarlinRenderingEngine (single-precision) in a follow-up issue ... to simplify the maintenance.

**Comment by bourgesl** [comment]
Sergey, is it good now ?
I ran jtreg: OK.

File: src/java.desktop/share/classes/sun/java2d/pipe/SpanShapeRenderer.java [hunk 6]
```diff
@@ -42,7 +42,6 @@
  * perform the actual rendering.
  */
 public abstract class SpanShapeRenderer implements ShapeDrawPipe {
-    static final RenderingEngine RenderEngine = RenderingEngine.getInstance();
```
Comment (prrace) [inline src/java.desktop/share/classes/sun/java2d/pipe/SpanShapeRenderer.java:45]: I suppose this is just clean up when you happened to notice it wasn't used/

File: test/jdk/sun/java2d/marlin/StrokedLinePerf.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import java.awt.geom.Line2D;
+import java.awt.image.BufferedImage;
+
+import static java.awt.image.BufferedImage.TYPE_INT_ARGB_PRE;
+
+/**
+ * @test
+ * @bug 7018932
+ * @summary fix LoopPipe.getStrokedSpans() performance (clipping enabled by Marlin renderer)
+ * @run main/othervm/timeout=10 -Dsun.java2d.renderer=sun.java2d.marlin.MarlinRenderingEngine StrokedLinePerf
```
Comment (prrace) [inline test/jdk/sun/java2d/marlin/StrokedLinePerf.java:37]: I hope 10 seconds is enough for a stable test .. not connecting to the windowing system so maybe.

File: src/java.desktop/share/classes/sun/java2d/pipe/SpanShapeRenderer.java (hunk 6 above)
Comment (bourgesl) [reply src/java.desktop/share/classes/sun/java2d/pipe/SpanShapeRenderer.java:45]: Exactly.

File: src/java.desktop/share/classes/sun/java2d/marlin/TransformingPathConsumer2D.java (hunk 4 above)
Comment (prrace) [inline src/java.desktop/share/classes/sun/java2d/marlin/TransformingPathConsumer2D.java:902]: are all the changes to double needed for this ? Or just preparation for removing the float pipeline ?

File: test/jdk/sun/java2d/marlin/StrokedLinePerf.java (hunk 7 above)
Comment (bourgesl) [reply test/jdk/sun/java2d/marlin/StrokedLinePerf.java:37]: Each render ops take up to 100ms max on my machine, so the test runs in 0.5s on my machine.

Would you prefer 30s ? 
Timeout is mandatory to detect test failure = hanging ...

File: src/java.desktop/share/classes/sun/java2d/marlin/TransformingPathConsumer2D.java (hunk 4 above)
Comment (bourgesl) [reply src/java.desktop/share/classes/sun/java2d/marlin/TransformingPathConsumer2D.java:902]: These change are just enough to make MarlinRenderingEngine (float) pass the new test.
Later we will propose to remove the float pipeline.

**Review by mrserb — APPROVED** [review]
Internal testing is fine, The fix looks good. Thank you for your contribution!

**Review by prrace — APPROVED** [review]
